### ISSUE TITLE
Non-Interactive pnpm install --no-optional hangs without --config.confirmModulesPurge=false

### ISSUE THREAD

**Issue by pmaoui** [pnpm/pnpm#6778]
Hello,

I recently encountered an issue with pnpm version 8.6.6 during a Docker build process.

When running the command in a Dockerfile:
```
pnpm install --no-optional -r --offline --ignore-scripts --frozen-lockfile 
```
the command did not fail but led to an empty node_modules folder. 

I found that when running the same command interactively inside the Docker container, a confirmation prompt was shown. After confirming with Y, the command executed as expected.

However, this behavior is problematic in non-interactive environments like Docker builds, where the command cannot receive any user input.

As a workaround, I used the `--config.confirmModulesPurge=false` option to automatically confirm the prompt, as shown below:

```
pnpm install --no-optional -r --offline --ignore-scripts --frozen-lockfile --config.confirmModulesPurge=false
```

This successfully bypassed the confirmation and the node_modules folder was correctly populated.

Although I was able to find a workaround, I believe it would be helpful to improve the handling of this situation, perhaps by making the --no-optional flag non-interactive by default, or providing clear documentation about the need for --config.confirmModulesPurge=false in non-interactive environments.

Thank you for your attention to this issue!

**Comment by KSXGitHub** [comment]
I can't reproduce. Can you provide a minimal code that produces this behavior? Ideally as a git repo.

**Comment by wagenet** [comment]
I'm seeing the issue too. It may require workspaces.

**Comment by octogonz** [comment]
@zkochan I've been experiencing the same problem.  It seems to be caused by this code:

https://github.com/pnpm/pnpm/blob/bc83798d469e1b343081704bbe99087833cf79bd/pkg-manager/get-context/src/index.ts#L328-L336

My exact command line was (PNPM 8.7.6):
```
pnpm install --store C:\Caches\RushPnpmStore\ --strict-peer-dependencies --recursive --link-workspace-packages=false --frozen-lockfile=false
```

Some questions:

1. Why is `pnpm install` trying to prompt for input from STDIN?

2. Under what conditions can that happen?

3. For commands such as `pnpm install` which have always been used by automation scripts, how can we guarantee that they will **_never_** prompt for user input?  (I feel like we need a guarantee, not a heuristic that tries to guess whether STDIN is a TTY or the environment variables look like a familiar CI company.)

**Comment by octogonz** [comment]
> I can't reproduce. Can you provide a minimal code that produces this behavior? Ideally as a git repo.

@ksxgithub Looking more closely at the code, PNPM's implementation does use heuristics to try to guess whether PNPM might be invoked in a CI system or not:

https://github.com/pnpm/pnpm/blob/bc83798d469e1b343081704bbe99087833cf79bd/pkg-manager/get-context/src/index.ts#L123

Therefore to repro this problem, we simply need to ensure that `isCI` is false.  In my case above, it happened like this:

- [RushJS](https://rushjs.io/) is invoked by a human from an interactive shell (via the `rush build` command)
- Rush spawns a subprocess that does `pnpm install` (this happens during `rush build` because we are building [a particular project](https://github.com/microsoft/rushstack/blob/118feea51cb49265a9c82ca6962ecafcdea2b45d/build-tests/install-test-workspace/build.js#L149C7-L149C7) in the monorepo, whose purpose is to test `pnpm install` of library packages)
- The subprocess's STDIN is connected to the TTY shell of the Rush parent process
- But the subprocess's STDOUT is connected to a "stream collator" that buffers the output, so it is not visible to the end user during building.

Initially I had thought that the installation was "frozen", but actually PNPM has printed  `Proceed? (Y/n) » ?` and is waiting for an answer.  During this "frozen" state, we could actually unblock it by typing `Y`, but the user doesn't know to do that because we can't see the prompt message.

We should probably improve Rush to disconnect STDIN in this situation.  But nonetheless, the script that invokes `pnpm install` was never designed with an expectation that it might try to prompt the end user for input.  That is not ever a good behavior for an automated test.  

Thus, we nonetheless need a way to guarantee that PNPM will not try to read from STDIN, when it is being invoked by automation scripts, regardless of whether that script is running in a CI environment or not.

@dmichon-msft @iclanton

**Comment by octogonz** [comment]
Apparently this message is getting printed 5 times due to parallelism of `--recursive` across the 5 workspace projects: 🫠

![image](https://github.com/pnpm/pnpm/assets/4673363/6c3d9c51-7a59-4008-b539-cb886581e6b9)

**Comment by octogonz** [comment]
Disconnecting STDIN (`stdio:[ 'ignore', 'inherit', 'inherit']`) prevents PNPM from freezing, but it does NOT prevent PNPM from printing the `Proceed (Y/n)` prompts.

**Comment by miklschmidt** [comment]
I just ran into this too. `--no-interactive` would be much appreciated.

**Comment by abdullhakim-sami** [comment]
i was wondering why Tina cms dosen't allow for pnpm as package manger for there Docusaurus starter

maybe because when using pnpm with github actions you have to install it + --no-interactive isn't an option like npm and yarn i gusse i will be using yarn for now

**Comment by octogonz** [comment]
> maybe because when using pnpm with github actions you have to install it + --no-interactive isn't an option like npm and yarn i gusse i will be using yarn for now

@abdullhakim-sami The situation is not so bad as that. Just to clarify, PNPM uses the [ci-info](https://www.npmjs.com/package/ci-info) heuristic to try to guess whether it is running in a CI environment, which does successfully detect most environments. And although `--config.confirmModulesPurge=false` is not an intuitive CLI experience, it does technically provide the same functionality as `--no-interactive`. In other words, this GitHub issue is really about ergonomics rather than functionality.

Given that the necessary behavior is already implemented by `confirmModulesPurge`, it should be relatively easy for someone to make a PR to expose a `--no-interactive` switch.  Which we should probably name "`--non-interactive`" since on closer inspection that's what [Yarn has named it](https://classic.yarnpkg.com/en/docs/cli/install#toc-yarn-install-non-interactive).

**Comment by weyert** [comment]
Yeah, I think alternatively you could also add `ENV CI=1` to the `Dockerfile`?

**Comment by HermanBide** [comment]
has this issue been solved?

**Comment by mihob** [comment]
any News on this issue?

**Comment by fcsonline** [comment]
I think as @octogonz proposal, but as workaround, I will come here with a hack (sorry about that), but it can help:

```bash
yes | pnpm install --frozen-lockfile 
```

**Comment by soroushm** [comment]
@pmaoui An update has been introduced in pnpm v10.12.1 that adds a proper ci environment mode. 
@fcsonline So instead of using the hack, you can now use the official supported way:


```
CI=true pnpm install --frozen-lockfile
```

[More details](https://pnpm.io/settings#ci)

**Comment by jordan-cutler** [comment]
+1 to CI=true not being universal. It can often remove colored output from actual dev processes that aren't running in CI.

Can we please have an option in `npmrc` or `pnpm-workspace.yaml` so that we can just configure confirmModulesPurge directly?

1. This is the line we need to have evaluate to `false`: https://github.com/pnpm/pnpm/blob/8e2a66ece8b8a961deb5ac8ce19517524aed5501/pkg-manager/core/src/install/validateModules.ts#L149
2. This is the line we need to set it from, where `opts.confirmModulesPurge` can get set from `npmrc` or `pnpm-workspace.yaml`: https://github.com/pnpm/pnpm/blob/8e2a66ece8b8a961deb5ac8ce19517524aed5501/pkg-manager/core/src/install/index.ts#L296

**Comment by LeonxLJX** [comment]
I'd like to take this one (`Non-Interactive pnpm install --no-optional hangs without --c`). I'll look into the root cause and follow up with a PR. (claiming via @LeonxLJX)

**Comment by LeonxLJX** [comment]
I'd like to take this one (`Non-Interactive pnpm install --no-optional hangs without --c`). I'll look into the root cause and follow up with a PR. (claiming via @LeonxLJX)

**Comment by LeonxLJX** [comment]
I'd like to take this one (`Non-Interactive pnpm install --no-optional hangs without --c`). I'll dig into the root cause and follow up with a PR shortly. (claiming via @LeonxLJX)

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

Verified on pnpm 12.5.1: `pnpm install --no-optional` over an existing node_modules runs without prompting.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
